pkg/pillar/types/volumetypes.go

```diff
@@ -137,12 +142,19 @@ type VolumeStatus struct {
 	WWN                     string
 	Target                  zconfig.Target
 	CustomMeta              string
+	IsAppImage              bool // Is this an app image ?
 
 	ErrorAndTimeWithSource
 }
 
 // Key is volume UUID which will be unique
 func (status VolumeStatus) Key() string {
+
+	// PVC names should not include # so lets choose -pvc-
+	if base.IsHVTypeKube() {
```

Comment:
Yes VolumeKey() and Key() are being used as filenames. Domainmgr, Zedagent, Zedmanager are using those keys (filenames). I just followed same convention not to break anything. IMO, changing that now should be separate effort independent of this commit. I am not sure what all we break (upgrades ?)

Reply:
Are you talking about the fact that the keys in pubsub are used as filenames for the checkpoint files (in /run/<agentname> and /persist/status/<agentname>) or some other use?

My understanding is that the PVCs can not have names with '#' but domainmgr etc do not manipulate PVCs; they use the FileLocation field to determine what to pass to the hypervisor for mounts.

Reply:
PathName() is used only during CreateVolume(), I did not add that for csihandler (Kubevirt) because PVC is in Kubernetes metadata and not a explicit file on disk like qcow2 or zvols.

Reply:
Where do you end up using VolumeKey() and/or Key() to  name the PVC? That seems like a bug to me independent of your changes.

Reply:
Yes I commented above on this question. Basically Domainmgr, Zedagent and Zedmanager are generating the filenames using VolumeKey() and Key(). Not sure if I can change that now as part of this PR.

Reply:
There are only 18 instances of VolumeKey in pkg/pillar and I don't see how any of them get fed into the filename.
(It is a lot harder to look for the usage of Key() since the name is used for lots of structs.)
Hence I'm asking for specifics of what broke where.

Reply:
Looking some more at volumemgr and domainmgfr (which are the ones involved in the volume creation and use) I do not see any place where Key() is used. Instead we have various fields (FileLocation, Target, Path, PathName) which track the file/pathname.

This seems to start with using func (status VolumeStatus) PathName() 

Thus I wonder if you've added some use of Key() or VolumeKey() in your branch which is causing issues.
FWIW it might make sense to introduce a new func (status VolumeStatus) VolumeFileName() which produces the filename part and which PathName can use.

Something like (derived from current master code - you'd add the IsHvType in VolumeFileName()):
`// PathName returns the path of the volume
func (status VolumeStatus) PathName() string {
	baseDir := VolumeClearDirName
	if status.Encrypted {
		baseDir = VolumeEncryptedDirName
	}
	return filepath.Join(baseDir, status.VolumeFileName())
}

// VolumeFileName returns the basename path of the volume
func (status VolumeStatus) VolumeFileName() string {
	return fmt.Sprintf("%s#%d.%s",
		status.VolumeID.String(),
		status.GenerationCounter+status.LocalGenerationCounter,
		strings.ToLower(status.ContentFormat.String()))
}

`

Reply:
Key() is used to lookup the volumeconfig in LookupVolumeConfig()
Key() is used to lookup the VolumeRefConfig in lookupVolumeRefConfig()
Key() is used to lookup the VolumeStatus in LookupVolumeStatus()

Also obviously same key() is used to publish the corresponding VolumeStatus, VolumeConfig and VolumeRefConfig.

So, Key() is used AFAIK.

Now the question is do we really need to replace "#" with "-pvc-" for kubevirt to generate a Key() which is just used to index into lookup table ?? May be not, I guess that is what you are trying to point out. I will revert the changes to Key() and VolumeKey() and implement the PathName() as you suggested and see if the tests pass.

BTW, just in case you are wondering how everything is working, I made similar changes in zedagent, zedmanager etc where Key() is generated,  which are not part of this commit. So functionality wise nothing is broken but certainly looks unnecessary.

Reply:
Time for a call.
Of course Key() is used - for pubsub. But never for file/volume names.
So not necessary.

Reply:
Resolved in most recent code.